File: Last_stone_weight 24-Aprill/Code/.Solution.java
import java.util.PriorityQueue;

import javax.swing.LayoutStyle;

import java.util.Comparator;
class Solution{
    public static int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Comparator.reverseOrder());
        for(int i=0;i<stones.length;i++){
            pq.add(stones[i]);
        }
        int p,q;
        while(pq.size()>1){
            p=pq.poll();
            q=pq.poll();
            // System.out.println(÷p+" "+q);
            pq.add(p-q);
        }
        return pq.poll();
        
    }
    public static void main(String[] args) {
          int [] arr = {2,7,4,1,8,1};
          System.out.println(lastStoneWeight(arr));
    }
}

File: Last_stone_weight 24-Aprill/Code/Solution.cpp
#include<bits/stdc++.h>
using namespace std;
//TIME AND SPACE COMPLEXITY
// TWO LOOPS ONE FOR PUSHIN ELEMENTS AND SECOND FOR DESTROYING THE STONES 
// O(N) + IN EVERY STEP EITHER A STONE IS DESTROYED OR BOTH STONE DESTROYED SO CLOSE TO O(N)
// OVERALL TIME COMPLEXITY O(N)
// S(N) USING PRIORITY QUEUE OF SIZE N
// PROBLEM LINK https://leetcode.com/problems/last-stone-weight/description/
class Solution {
public:    
   int lastStoneWeight(vector<int>& stones) {
        int n = stones.size();
        if(n == 1) return stones[0];
        priority_queue<int>pq;
        for(auto it:stones){
            pq.push(it);
        }
        while(pq.size() > 1){
            int x,y;
            y = pq.top();
            pq.pop();
            x = pq.top();
            pq.pop();
           if(x != y){
               y = y - x;
               pq.push(y);
           }
        }
        if(pq.size() == 0) return 0;
        else return pq.top();
    }
};
int main(){
Solution obj;
vector<int>arr = {2,7,4,1,8,1};
cout<<obj.lastStoneWeight(arr)<<endl;
return 0;
}